text/0162-refactoring-support.md

```diff
@@ -0,0 +1,829 @@
+feature.
+
+For both `deploy` and `refactor`:
+
+- `--refactor-mapping=<FILE>`: use the mapping from a file, instead of computing
+  it.
+- `--skip-refactoring=<FILE>`: removes the elements listed in the file from the
+  computed mapping.
+- `--unstable=refactor`: used to acknowledge that this feature is experimental.
+  If the flag is not set, and the CLI would try to perform some refactor, the
+  command fails with an error message explaining why.
+
+For `deploy` only:
+
+- `--refactoring-action=[ACTION]`: the action to take in case there is a
+  refactor to be made (either computed automatically or provided via a mapping
+  file). Possible values for `ACTION` are:
+    - `refactor`: automatically refactor and deploy.
+    - `skip`: deploy without refactoring. Default value on non-interactive mode.
+    - `fail`: exit with non-zero status code.
+    - `ask`: ask the user what to do. Default value on interactive mode.
+
+For `refactor` only:
+
+- `--dry-run`: print the mapping to the console, but do not apply it.
+- `--force`: go ahead with refactoring, without prompting the user.
+
+All these settings are also available in the `cdk.json` file:
+
+```json
+{
+  "app": "...",
+  "refactor": {
+    "refactoringAction": "refactor",
+    "dryRun": true
+  }
+}
+```
+
+### Rollback
+
+During the execution of a `deploy` command, the CLI will do the refactor and
+then proceed to the actual deployment (assuming the right set of flags). If the
+deployment fails, and CloudFormation rolls it back, the CLI will execute a
+second refactor, in reverse, to bring the resources back to their original
+locations.
+
+If you don't want the CLI to perform the rollback refactor, you can use the
+`--no-rollback` flag, which also controls the rollback behavior of the
+deployment.
+
+### Programmatic access
+
+The same refactoring feature is also available in the CDK toolkit library:
+
+```typescript
+declare const toolkit: Toolkit;
+declare const cx: ICloudAssemblySource;
+
+// To execute possible refactors as part of the deploy operation:
+await toolkit.deploy(cx, {
+  refactoring: RefactoringMode.EXECUTE_AND_DEPLOY
+});
+
+// Or, if you just want to refactor the stacks:
+await toolkit.refactor(cx);
+```
+
+### Skipping classes of constructs
+
+In the CDK construct library there are some constructs that rely on
+CloudFormation's behavior to force the replacement of resources. API Gateway's
+`Deployment` and Lambda's `Version` are two examples of this strategy. By
+default, the CLI will detect these changes as refactors and not replacements. We
+have already seen a mechanism to override this: skip files. But it would be too
+inconvenient to have to maintain skip files and remembering to update it every
+time the logical ID changes.
+
+To solve this, CDK constructs can be automatically excluded by calling the new
+method `Stack.skipRefactoring(constructToBeSkipped)`. By calling this method,
```

Comment:
@rix0rrr is this a good way to do it?

Reply:
I think it will be some form of metadata.

```ts
Annotations.of(constructToBeSkipped).add(SKIP_REFACTORING, true);
```

Or something like that? Could be abstracted via a `Stack` method, sure.